src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -10204,17 +10212,51 @@ static bool isCandidateUnique(ref MethodSymbol? method, MethodSymbol candidate)
 
             if (node.SearchExtensionMethods)
             {
-                var receiver = node.ReceiverOpt!;
+                var receiver = node.ReceiverOpt;
+                Debug.Assert(receiver is not null);
+
                 var methodGroup = MethodGroup.GetInstance();
                 foreach (var scope in new ExtensionScopes(this))
                 {
+                    methodGroup.Clear();
+                    populateExtensionTypeMethodsFromSingleBinder(scope, methodGroup, node.Syntax, receiver, node.Name, typeArguments);
+                    foreach (MethodSymbol method in methodGroup.Methods)
+                    {
+                        var substituted = typeArguments.IsDefaultOrEmpty ? method : method.Construct(typeArguments);
+                        if (!satisfiesConstraintChecks(substituted))
+                        {
+                            continue;
+                        }
+
+                        var wasUnique = isCandidateUnique(ref foundMethod, substituted);
+                        if (!wasUnique)
+                        {
+                            methodGroup.Free();
+                            return null;
+                        }
+                    }
+
+                    // PROTOTYPE we'll want to revisit to merge extension methods and methods from extension types within the same scope
+                    if (foundMethod is not null)
+                    {
+                        methodGroup.Free();
+                        return foundMethod;
+                    }
+
                     methodGroup.Clear();
                     PopulateExtensionMethodsFromSingleBinder(scope, methodGroup, node.Syntax, receiver, node.Name, typeArguments, BindingDiagnosticBag.Discarded);
                     foreach (var extensionMethod in methodGroup.Methods)
                     {
                         var substituted = typeArguments.IsDefaultOrEmpty ? extensionMethod : extensionMethod.Construct(typeArguments);
 
-                        var reduced = substituted.ReduceExtensionMethod(receiver.Type, Compilation, out bool wasFullyInferred);
+                        var type = receiver.Type;
+                        Debug.Assert(type is not null);
+                        if (type.ExtendedTypeNoUseSiteDiagnostics is { } extendedType)
+                        {
+                            type = extendedType;
```

Comment:
>type = extendedType; [](http://example.com/codeflow?start=28&length=20)

Why is this the right thing to do? What if the extension method extends the extension type? What if it extends a type parameter?

Reply:
> What if the extension method extends the extension type?

That's not allowed:
> We modify the extension methods rules as follows:
> [...] The first parameter of an extension method may have no modifiers other than this, and the parameter type may not be a pointer or an extension type.

> What if it [the extension method] extends a type parameter?

I think that's only possible in error scenarios, as extension methods must be defined in a top-level non-generic class.

Reply:
>> What if it [the extension method] extends a type parameter?

> I think that's only possible in error scenarios, as extension methods must be defined in a top-level non-generic class.


For example, this isn't an error scenario:
```
static class Ext
{
    public static T M<T>(this T a) => a;
}
```

Reply:
> We modify the extension methods rules as follows

Consider capturing the fact in comments, to make motivation for the logic clear.